packages/core/bootstrap/src/lib/modules/requester.ts

```diff
@@ -34,17 +34,43 @@ import type {
   BatchableProperty,
 } from '../../types'
 
-type CustomError<T = unknown> = (data: T) => boolean
+export type CustomErrorReturnValue = boolean | string
+type CustomError<T = unknown> = (data: T) => CustomErrorReturnValue
 const defaultCustomError = () => false
 
 export class Requester {
+  // used to return consistent error messages when failing with or without a customError
+  static generateErrorMessage<T>(
+    config: AxiosRequestConfig,
+    response?: AxiosResponse<T>,
+    append?: string,
+    customErrorResult?: CustomErrorReturnValue,
+  ): string {
+    const { baseURL, url, data, params } = config // strip noisy and sensitive data from axios config before logging/including in the response
+
+    const preamble = customErrorResult
```

Comment:
Should we also check `typeof customErrorResult === 'string'` here?

Reply:
ok cool, I think I understand.

However, unrelated but still relevant to this line - shouldn't we return `Could not retrieve valid data from Data Provider [...]` if `customErrorResult` is set? Now it seems like we return `Caught error trying to fetch data from Data Provider.` on `customErrorResult`.